File: packages/sdk/hush-py/src/clawdstrike/guards/patch_integrity.py
"""Patch integrity guard - validates code patches."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Optional, Tuple

from clawdstrike.guards.base import Guard, GuardAction, GuardContext, GuardResult, Severity


@dataclass
class PatchIntegrityConfig:
    """Configuration for PatchIntegrityGuard."""

    max_additions: int = 1000
    max_deletions: int = 500
    require_balance: bool = False
    max_imbalance_ratio: float = 5.0
    forbidden_patterns: List[str] = field(default_factory=list)


class PatchIntegrityGuard(Guard):
    """Guard that validates patch size, balance, and forbidden patterns."""

    def __init__(self, config: Optional[PatchIntegrityConfig] = None) -> None:
        self._config = config or PatchIntegrityConfig()
        self._compiled_forbidden: List[tuple[str, re.Pattern[str]]] = []
        for pattern_str in self._config.forbidden_patterns:
            try:
                compiled = re.compile(pattern_str)
            except re.error as e:
                raise ValueError(
                    f"Invalid regex in forbidden_patterns: {pattern_str!r}: {e}"
                ) from e
            self._compiled_forbidden.append((pattern_str, compiled))

    @property
    def name(self) -> str:
        return "patch_integrity"

    def handles(self, action: GuardAction) -> bool:
        return action.action_type == "patch"

    def _count_changes(self, diff: str) -> Tuple[int, int]:
        """Count additions and deletions in a diff."""
        additions = 0
        deletions = 0

        for line in diff.split("\n"):
            if line.startswith("@@") or line.startswith("---") or line.startswith("+++"):
                continue
            if line.startswith("+") and not line.startswith("+++"):
                additions += 1
            elif line.startswith("-") and not line.startswith("---"):
                deletions += 1

        return additions, deletions

    def check(self, action: GuardAction, context: GuardContext) -> GuardResult:
        if not self.handles(action):
            return GuardResult.allow(self.name)

        diff = action.diff
        if diff is None:
            return GuardResult.allow(self.name)

        additions, deletions = self._count_changes(diff)

        # Check additions limit
        if additions > self._config.max_additions:
            return GuardResult.block(
                self.name,
                Severity.ERROR,
                f"Patch exceeds max additions: {additions} > {self._config.max_additions}",
            ).with_details({
                "additions": additions,
                "deletions": deletions,
                "max_additions": self._config.max_additions,
            })

        # Check deletions limit
        if deletions > self._config.max_deletions:
            return GuardResult.block(
                self.name,
                Severity.ERROR,
                f"Patch exceeds max deletions: {deletions} > {self._config.max_deletions}",
            ).with_details({
                "additions": additions,
                "deletions": deletions,
                "max_deletions": self._config.max_deletions,
            })

        # Check balance if required
        if self._config.require_balance and deletions > 0:
            ratio = additions / deletions
            if ratio > self._config.max_imbalance_ratio:
                return GuardResult.block(
                    self.name,
                    Severity.WARNING,
                    f"Patch imbalance ratio too high: {ratio:.1f} > {self._config.max_imbalance_ratio}",
                ).with_details({
                    "additions": additions,
                    "deletions": deletions,
                    "ratio": ratio,
                    "max_ratio": self._config.max_imbalance_ratio,
                })

        # Check forbidden patterns in the diff
        for pattern_str, compiled in self._compiled_forbidden:
            match = compiled.search(diff)
            if match:
                return GuardResult.block(
                    self.name,
                    Severity.CRITICAL,
                    f"Patch contains forbidden pattern: {pattern_str}",
                ).with_details({
                    "forbidden_pattern": pattern_str,
                    "additions": additions,
                    "deletions": deletions,
                })

        return GuardResult.allow(self.name).with_details({
            "additions": additions,
            "deletions": deletions,
        })


__all__ = ["PatchIntegrityGuard", "PatchIntegrityConfig"]

File: packages/sdk/hush-ts/src/guards/patch-integrity.ts
import { Guard, GuardAction, GuardContext, GuardResult, Severity } from "./types";

const DEFAULT_FORBIDDEN_PATTERNS = [
  // Disable security features
  /disable[ _\-]?(security|auth|ssl|tls)/i,
  /skip[ _\-]?(verify|validation|check)/i,
  // Dangerous operations
  /rm\s+-rf\s+\//i,
  /chmod\s+777/i,
  /eval\s*\(/i,
  /exec\s*\(/i,
  // Backdoor indicators
  /reverse[_\-]?shell/i,
  /bind[_\-]?shell/i,
  /base64[_\-]?decode.*exec/i,
];

export interface PatchIntegrityConfig {
  /** Enable/disable this guard */
  enabled?: boolean;
  /** Maximum lines added in a single patch */
  maxAdditions?: number;
  /** Maximum lines deleted in a single patch */
  maxDeletions?: number;
  /** Regex patterns that are forbidden in patches */
  forbiddenPatterns?: RegExp[];
  /** Require patches to have balanced additions/deletions */
  requireBalance?: boolean;
  /** Maximum imbalance ratio (additions/deletions) */
  maxImbalanceRatio?: number;
}

export interface PatchAnalysis {
  additions: number;
  deletions: number;
  imbalanceRatio: number;
  forbiddenMatches: ForbiddenMatch[];
  exceedsMaxAdditions: boolean;
  exceedsMaxDeletions: boolean;
  exceedsImbalance: boolean;
}

export interface ForbiddenMatch {
  line: string;
  pattern: string;
}

/**
 * Guard that validates patch safety.
 *
 * Checks for:
 * - Forbidden patterns (security disabling, dangerous commands, backdoors)
 * - Size limits (max additions/deletions)
 * - Balance (additions vs deletions ratio)
 */
export class PatchIntegrityGuard implements Guard {
  readonly name = "patch_integrity";
  private enabled: boolean;
  private maxAdditions: number;
  private maxDeletions: number;
  private forbiddenPatterns: RegExp[];
  private requireBalance: boolean;
  private maxImbalanceRatio: number;

  constructor(config: PatchIntegrityConfig = {}) {
    this.enabled = config.enabled ?? true;
    this.maxAdditions = config.maxAdditions ?? 1000;
    this.maxDeletions = config.maxDeletions ?? 500;
    this.forbiddenPatterns = config.forbiddenPatterns ?? DEFAULT_FORBIDDEN_PATTERNS;
    this.requireBalance = config.requireBalance ?? false;
    this.maxImbalanceRatio = config.maxImbalanceRatio ?? 10.0;
  }

  handles(action: GuardAction): boolean {
    return this.enabled && action.actionType === "patch";
  }

  check(action: GuardAction, _context: GuardContext): GuardResult {
    if (!this.enabled) {
      return GuardResult.allow(this.name);
    }
    if (!this.handles(action)) {
      return GuardResult.allow(this.name);
    }

    const diff = action.diff;
    if (!diff) {
      return GuardResult.allow(this.name);
    }

    const analysis = this.analyze(diff);

    if (this.isSafe(analysis)) {
      return GuardResult.allow(this.name);
    }

    const issues: string[] = [];

    if (analysis.forbiddenMatches.length > 0) {
      const patterns = analysis.forbiddenMatches.map((m) => m.pattern).join(", ");
      issues.push(`Contains forbidden patterns: ${patterns}`);
    }

    if (analysis.exceedsMaxAdditions) {
      issues.push(`Too many additions: ${analysis.additions} (max: ${this.maxAdditions})`);
    }

    if (analysis.exceedsMaxDeletions) {
      issues.push(`Too many deletions: ${analysis.deletions} (max: ${this.maxDeletions})`);
    }

    if (analysis.exceedsImbalance) {
      issues.push(
        `Imbalanced patch: ratio ${analysis.imbalanceRatio.toFixed(2)} (max: ${this.maxImbalanceRatio.toFixed(2)})`,
      );
    }

    const severity = analysis.forbiddenMatches.length > 0 ? Severity.CRITICAL : Severity.ERROR;

    return GuardResult.block(this.name, severity, issues.join("; ")).withDetails({
      path: action.path,
      additions: analysis.additions,
      deletions: analysis.deletions,
      imbalanceRatio: analysis.imbalanceRatio,
      forbiddenMatches: analysis.forbiddenMatches.length,
    });
  }

  /**
   * Analyze a unified diff.
   */
  analyze(diff: string): PatchAnalysis {
    let additions = 0;
    let deletions = 0;
    const forbiddenMatches: ForbiddenMatch[] = [];

    for (const line of diff.split("\n")) {
      if (line.startsWith("+") && !line.startsWith("+++")) {
        additions++;

        // Check for forbidden patterns in added lines
        for (const pattern of this.forbiddenPatterns) {
          if (pattern.test(line)) {
            forbiddenMatches.push({
              line,
              pattern: pattern.source,
            });
          }
        }
      } else if (line.startsWith("-") && !line.startsWith("---")) {
        deletions++;
      }
    }

    const imbalanceRatio = deletions > 0 ? additions / deletions : additions > 0 ? Infinity : 1.0;

    return {
      additions,
      deletions,
      imbalanceRatio,
      forbiddenMatches,
      exceedsMaxAdditions: additions > this.maxAdditions,
      exceedsMaxDeletions: deletions > this.maxDeletions,
      exceedsImbalance: this.requireBalance && imbalanceRatio > this.maxImbalanceRatio,
    };
  }

  private isSafe(analysis: PatchAnalysis): boolean {
    return (
      analysis.forbiddenMatches.length === 0 &&
      !analysis.exceedsMaxAdditions &&
      !analysis.exceedsMaxDeletions &&
      !analysis.exceedsImbalance
    );
  }
}
